tinygrad/renderer/assembly.py

```diff
@@ -0,0 +1,181 @@
+        cast(tmp, dtypes.half, dtypes.float32, u=u)
+      elif args == BinaryOps.CMPLT:
+        kernel.extend([lang.asm_for_op[args](pred:=ssa(None,'lt','pred'), *[r[x] for x in vin], lang.dtype_to_asmtype[vin[0].dtype]),
+                       f"selp.{lang.dtype_to_asmtype[dtype]} {ssa(u, 'alu')}, {const(1, dtype)}, {const(0, dtype)}, {pred};"])
+      elif args == TernaryOps.WHERE:
+        kernel.extend([f"setp.ne.{lang.dtype_to_asmtype[vin[0].dtype]} {(pred:=ssa(None,'wh','pred'))}, {r[vin[0]]}, {const(0, vin[0].dtype)};",
+                       lang.asm_for_op[args](ssa(u, "alu"), pred, r[vin[1]], r[vin[2]], lang.dtype_to_asmtype[dtype])])
+      else: kernel.append(lang.asm_for_op[args](ssa(u, "alu"), *[r[x] for x in vin], lang.dtype_to_asmtype[dtype]))
+    elif uop == UOps.DEFINE_ACC:
+      assert dtype is not None
+      kernel.append(f"mov.b{lang.dtype_to_asmtype[dtype][1:]} {ssa(u, 'acc')}, {const(args, dtype)};")
+    elif uop == UOps.SPECIAL:
+      if args[1].startswith("i"): kernel.extend([f"mov.u32 %{args[1]}, {lang.gid[args[0]]};", f"mov.u32 {(gdim:=ssa(None,'tmp','u32'))}, {lang.gdim[args[0]]};",
+                                                 f"mov.u32 {(lid:=ssa(None,'tmp','u32'))}, {lang.lid[args[0]]};", f"mad.lo.u32 %{args[1]}, %{args[1]}, {gdim}, {lid};"])
+      else: kernel.append(f"mov.u32 %{args[1]}, {(lang.gid if args[1].startswith('g') else lang.lid)[args[0]]};")
+      if args[1].startswith("l"): local_size.append(args[2])
+      r[u] = "%" + args[1]
+      kernel = [f".reg .u32 %{args[1]};"] + kernel
+    elif uop == UOps.CONST: r[u] = const(args, dtype, force_mov=True)
+    elif uop == UOps.LOAD:
+      assert dtype is not None and vin[1].dtype is not None
+      if vin[1] and vin[1].uop != UOps.CONST:
+        kernel.append(f"mad.wide.u32 {(loc:=ssa(None,'loc','u64'))}, {cast(r[vin[1]], dtypes.uint32, vin[1].dtype)}, {const(dtype.itemsize, dtypes.uint32, force_mov=True)}, {r[vin[0]]};")
+        loc = f"[{loc}]"
+      else: loc = f"[{r[vin[0]]}{f'+{int(vin[1].arg * dtype.itemsize)}' if vin[1] else ''}]"
+      val = ssa(u, 'val')
+      if dtype.itemsize == 1: tmp = ssa(None, 'tmp', 's8')
+      if len(vin) > 3:
+        assert vin[2].dtype is not None
+        kernel.append(f"setp.ne.{lang.dtype_to_asmtype[vin[2].dtype]} {(pred:=ssa(None,'ld','pred'))}, {r[vin[2]]}, {const(0, vin[2].dtype)};")
+      kernel.append(f"{f'@{pred} ' if len(vin) > 3 else ''}ld{'.shared' if vin[0].uop == UOps.DEFINE_LOCAL else ''}.{'s8' if dtype.itemsize == 1 else 'b16' if dtype == dtypes.float16 else lang.dtype_to_asmtype[dtype]} {tmp if dtype.itemsize == 1 else val}, {loc};")
+      if len(vin) > 3: kernel.append(f"@!{pred} mov.b{'8' if dtype.itemsize == 1 else lang.dtype_to_asmtype[dtype][1:]} {tmp if dtype.itemsize == 1 else val}, {r[vin[3]]};")
+      if dtype.itemsize == 1: kernel.append(f"cvt.{lang.dtype_to_asmtype[dtype]}.s8 {val}, {tmp};")
+    elif uop == UOps.PHI:
+      assert dtype is not None
```

Comment:
Why is this assert dtype line repeated over and over?

Reply:
Unfortunately without this, mypy complains about indexing into `dtype_to_asmtype` with a possibly None value, I suppose adding a dummy None entry to that dictionary could make the asserts not required?